src/objective-c/GRPCClient/GRPCCall.m

```diff
@@ -52,6 +55,205 @@ @interface GRPCCall ()<GRXWriteable>
+    }
+
+    self->_call = [[GRPCCall alloc] initWithHost:self->_requestOptions.host
+                                            path:self->_requestOptions.path
+                                      callSafety:self->_requestOptions.safety
+                                  requestsWriter:self->_pipe
+                                     callOptions:self->_callOptions];
+    if (self->_callOptions.initialMetadata) {
+      [self->_call.requestHeaders addEntriesFromDictionary:self->_callOptions.initialMetadata];
+    }
+    id<GRXWriteable> responseWriteable = [[GRXWriteable alloc] initWithValueHandler:^(id value) {
+      dispatch_async(self->_dispatchQueue, ^{
+        if (self->_handler) {
+          NSDictionary *headers = nil;
+          if (!self->_initialMetadataPublished) {
+            headers = self->_call.responseHeaders;
+            self->_initialMetadataPublished = YES;
+          }
+          if (headers) {
+            [self issueInitialMetadata:headers];
+          }
+          if (value) {
+            [self issueMessage:value];
+          }
+        }
+      });
+    }
+        completionHandler:^(NSError *errorOrNil) {
+          dispatch_async(self->_dispatchQueue, ^{
+            if (self->_handler) {
+              NSDictionary *headers = nil;
+              if (!self->_initialMetadataPublished) {
+                headers = self->_call.responseHeaders;
+                self->_initialMetadataPublished = YES;
+              }
+              if (headers) {
+                [self issueInitialMetadata:headers];
+              }
+              [self issueClosedWithTrailingMetadata:self->_call.responseTrailers error:errorOrNil];
+
+              // Clean up _handler so that no more responses are reported to the handler.
+              self->_handler = nil;
+
+              if (self->_call) {
+                [self->_pipe writesFinishedWithError:nil];
+                self->_call = nil;
+                self->_pipe = nil;
+              }
+            }
+          });
+        }];
+    [self->_call startWithWriteable:responseWriteable];
+  });
+}
+
+- (void)cancel {
+  dispatch_async(_dispatchQueue, ^{
+    if (self->_call) {
+      [self->_call cancel];
+      self->_call = nil;
+      self->_pipe = nil;
+    }
+    if (self->_handler) {
+      id<GRPCResponseHandler> handler = self->_handler;
+      dispatch_async(handler.dispatchQueue, ^{
+        if ([handler respondsToSelector:@selector(closedWithTrailingMetadata:error:)]) {
+          [handler closedWithTrailingMetadata:nil
+                                        error:[NSError errorWithDomain:kGRPCErrorDomain
+                                                                  code:GRPCErrorCodeCancelled
+                                                              userInfo:@{
+                                                                         NSLocalizedDescriptionKey :
+                                                                           @"Canceled by app"
```

Comment:
A bit tricky since `NSDebugDescriptionErrorKey` has been used for more verbose debug message... I'm thinking of two options: leave it as is (i don't know how bad it would be; people have been using it and are fine so far) or concatenate it in `NSDebugDescriptionErrorKey` (might confuse people as they do not find `NSDebugDescriptionErrorKey` any more) and I tend to leave it.